input_output.ipynb

```diff
@@ -0,0 +1,917 @@
+    "- reading and writing data to a file\n",
+    "- printing text output\n",
+    "- asking for user input\n",
+    "- connecting to databases or other network services\n",
+    "\n",
+    "\n",
+    "The majority of these operations are covered by the python standard library. We are going to see how to use them in this chapter.\n",
+    "\n",
+    "<div class=\"alert alert-block alert-info\">\n",
+    "⚠️ In reference to the chapter on functional programming, it is interesting to note that these functions perform *side-effects*.  Therefore any code containing these operations is no longer *pure* and is not referentially transparent. The same function can return different values for the same argument if called multiple time and the function can have *long-distance* effects. That means that they can modify the state of the program somewhere else leading to unexpected results. \n",
+    "    \n",
+    "Therefore, I always suggest trying to separate input and output from the other computations in your program. For example, if you have a complex calculation requring several user inputs at several stages of the process, consider writing a function that only performs the calculation given all inputs and then require all inputs separately, for example through a single file. This makes your code easier to debug, test and understand.\n",
+    "</div>\n",
+    "\n",
+    "\n",
+    "\n",
+    "\n",
+    "## String input and output \n",
+    "The most basic I/O operation in python is displaying text on the python console. This is done using the `print` function:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(\"I am some text\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The complemenentary operation, prompting for *user input* is done through `input`.\n",
+    "The function takes one string as argument, which gives the prompt displayed on the console. The return value of the function contains the input the user types:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "user_input = input(\"Write some text\")\n",
+    "print(f\"The user wrote {user_input}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "<div class=\"alert alert-block alert-info\">\n",
+    "⚠️ Note that `input` is a **blocking operation**. The program evaluation stops and waits until the user provides an input. Therefore, use `input` **very carefully* and only when truly necessary. A common issue is that someone first writes an interactive program expecting user input through `input` and later integrates it into a larger application that is supposed to run automatically without any user interaction. Suddenly, the application stops somehwere and does not run further because of a well hidden `input` call...\n",
```

Comment:
```suggestion
    "⚠️ Note that `input` is a **blocking operation**. The program evaluation stops and waits until the user provides input. Therefore, use `input` <b>carefully</b> and only when necessary. A common issue is that someone first writes an interactive program expecting user input through `input` and later integrates it into a larger application that is supposed to run automatically without user interaction. Suddenly, the application stops somewhere and does not run further because of a well-hidden `input` call...\n",
```